elf_reader.go

```diff
@@ -821,6 +843,62 @@ func resolveBTFArrayMacro(typ btf.Type) (btf.Type, error) {
 	return ptr.Target, nil
 }
 
+// resolveBTFValuesContents resolves relocations into ELF sections belonging
+// to btf.VarSecinfo's. This can be used on the 'values' member in BTF map
+// definitions to extract static declarations of map contents.
+func resolveBTFValuesContents(es *elfSection, vs *btf.VarSecinfo, member btf.Member) ([]MapKV, error) {
+	// The elements of a .values pointer array are not encoded in BTF.
+	// Instead, relocations are generated into each array index.
+	// However, it's possible to leave certain array indices empty, so all
+	// indices' offsets need to be checked for emitted relocations.
+
+	// The offset of the 'values' member within the _struct_ (in bits)
+	// is the starting point of the array. Convert to bytes. Add VarSecinfo
+	// offset to get the absolute position in the ELF blob.
+	start := (member.Offset / 8) + vs.Offset
+	// 'values' is encoded in BTF as a zero (variable) length struct
+	// member, and its contents run until the end of the VarSecinfo.
+	// Add VarSecinfo offset to get the absolute position in the ELF blob.
+	end := vs.Size + vs.Offset
+	// The size of an address in this section. This determines the width of
+	// an index in the array.
+	align := uint32(es.SectionHeader.Addralign)
+
+	// Check if variable-length section is aligned.
+	if (end-start)%align != 0 {
+		return nil, errors.New("unaligned static values section")
+	}
+	elems := (end - start) / align
+
+	if elems == 0 {
+		return nil, nil
+	}
+
+	contents := make([]MapKV, 0, elems)
+
+	// k is the array index, off is its corresponding ELF section offset.
+	for k, off := uint32(0), start; k < elems; k, off = k+1, off+align {
+		r, ok := es.relocations[uint64(off)]
+		if ok {
+			// Relocation exists for the current offset in the ELF section.
+			// Emit a value stub based on the type of relocation to be replaced by
+			// a real fd later in the pipeline before populating the map.
+			// Map keys are encoded in MapKV entries, so empty array indices are
+			// skipped here.
+			switch t := elf.ST_TYPE(r.Info); t {
+			case elf.STT_FUNC:
```

Comment:
What if we check elfCode.sections[r.Section???].kind instead? The STT_FUNC / STT_OBJECT semantics are kind of brittle, the less we rely on it the better I think.

Reply:
I did some more digging and found that relocations also have their 'source' section index attached, the section where the symbol itself lives. In our test programs, the section of relocations of type `STT_FUNC` points to wherever the function lives. Relocations of type `STT_OBJECT` point to the `.maps` section index.

```
λ  ebpf prog-array-autopop ✗ readelf -Sgt -x .maps -r testdata/btf_map_init-el.elf
There are 26 section headers, starting at offset 0x10f8:

Section Headers:
  [Nr] Name
       Type              Address          Offset            Link
       Size              EntSize          Info              Align
       Flags
...
  [ 3] socket/tail
       PROGBITS         0000000000000000  0000000000000040  0
       0000000000000010 0000000000000000  0                 8
       [0000000000000006]: ALLOC, EXEC
...
  [ 6] .maps
       PROGBITS         0000000000000000  0000000000000080  0
       0000000000000078 0000000000000000  0                 8
       [0000000000000003]: WRITE, ALLOC
```

Unfortunately, both the program and .maps sections are of type `PROGBITS`. `.maps` has different flags than `socket/tail` in this example, but going off of that doesn't feel any less brittle than using the relocation symbol type to begin with. Such seems to be the nature of ELF parsing.

Is this what you had in mind? Any other suggestions?

Reply:
I mean using this info: https://github.com/cilium/ebpf/blob/master/elf_reader.go#L219-L231 The loader already classifies sections based on some heuristics. Instead of asking "is the symbol of the correct type" we'd check that the symbol's section has the correct kind. Writing this out, it's probably best to stick with what you have right now. We'll return an error if the map or program don't exist when creating the map.